doc/developer/design/20231027_refresh_mvs.md

```diff
@@ -0,0 +1,276 @@
+
+### Implementation Alternative: Suspending the Replica Process to Disk
+
+As mentioned in the [scoping section](#out-of-scope), an alternative implementation that would cover also intermediate refresh intervals would be to suspend the replica process and save its entire state to S3 and a local disk cache, and restore it when a refresh is needed. [There are existing techniques for saving process state, e.g., CRIU.](https://faun.pub/kubernetes-checkpointing-a-definitive-guide-33dd1a0310f6) However, this would be a much bigger project than the proposed solution.
+
+## Automated Cluster Scheduling for `REFRESH EVERY` MVs
+
+TODO: updated discussion
+
+## Introspection / Observability
+
+For showing the `REFRESH` options of each materialized view, I'm planning to create a new table in `mz_internal` called `mz_materialized_view_refresh_options`, which will have a row for each non-trivial refresh option of each MV. The columns will be `(materialized_view_id text, interval interval, interval_aligned_to timestamptz, at timestamptz)`, and only those columns will have a non-null value that are relevant for the refresh option that the row is about (e.g., a `REFRESH EVERY` would be non-null in interval and interval_aligned_to). `REFRESH ON COMMIT` is considered a trivial refresh option, and won't show up in the table. (See Nikhil's [msg](https://materializeinc.slack.com/archives/C06535JL58R/p1701228087745239?thread_ts=1700595291.424189&cid=C06535JL58R) on an old design thread.)
+
+For showing the previous and next refresh of each MV, I'm planning to create a new table in `mz_internal` called `mz_materialized_view_refreshes (materialized_view_id text, last_completed_refresh timestamptz, next_refresh timestamptz)`. (The values would be derived from the write frontier of the MV's storage collection.)
+
+For seeing whether the last refresh is being late, the user can run `EXPLAIN TIMESTAMP` in `STRICT SERIALIZABLE` mode, and look at can respond immediately. If it's false, then the last refresh's completion is overdue. Another way to get the same information would be to check if `mz_materialized_view_refreshes.next_refresh < now()`.
+
+For showing cluster schedules, I'm thinking of doing a similar thing as in `mz_materialized_view_refresh_options`, because eventually the `SCHEDULE =` option will allow for specifying multiple schedules. So I'd create a new table in `mz_internal` called `mz_cluster_schedules`, and it would have a row for each non-trivial (non-`MANUAL`) schedule option of each cluster. This would currently be only `SCHEDULE = ON REFRESH`. Columns would be `(cluster_id text, refresh_rehydration_time_estimate interval)`. (Eventually, we'll probably also want a `next_scheduled_turn_on`, but this doesn't seem so urgent. It will get more important when we'll be choosing the warmup time automatically.)
```

Comment:
> `refresh_rehydration_time_estimate interval`

I haven't been keeping up with what a these schedules are so I don't know what this is, but it's a bit of a smell that the column name has the word "time" but the type is `interval` not `time`.

Reply:
It specifies how much earlier we should turn on the cluster than the refresh (so that it can rehydrate before the refresh is due). The corresponding syntax is
```
SCHEDULE = ON REFRESH (REHYDRATION TIME ESTIMATE = '1 hour')
```
There was a ([SQL council discussion](https://materializeinc.slack.com/archives/C063H5S7NKE/p1712270012125849?thread_ts=1712165305.916299&cid=C063H5S7NKE)) when we named this concept. I think we avoided using the word "interval" to prevent confusion with the refresh interval.

Reply:
Yes, it's already going out to [the user](https://github.com/MaterializeInc/accounts/issues/3) next Thursday, so I'd like to avoid changing it at this point, unless absolutely necessary.

Reply:
Ah OK, in that case the name is fine.